### PR TITLE
IDCOM-2129 Extend Transaction

### PR DESCRIPTION
**Pull request by dankelleher**
Extend Transaction:

This PR adds the ability to extend a transaction once proposed, allowing much larger transactions to be executed via cryptid.

Note: Execution still happens in one transaction, so the compute budget still applies. 

Also: this PR does not include any functionality for the client to deduce if a transaction needs to be split up into multiple extend() steps. This is left to the integrator at present. Later we may want to add this.

The PR adds two commands to the cryptid program:

- extend: add instructions to the cryptid transaction
- seal: Mark the transaction as ready to execute

It also adds the ability to propose a transaction in NotReady state. (Default is Ready)

This PR also includes some Readme for the client sdk, and tests which give a bunch of examples, but the tl:dr is:

```
// Propose an initial transaction
const {proposeTransaction, proposeSigners, transactionAccount} = await cryptid.propose(tx1, TransactionState.NotReady);
await cryptid.send(proposeTransaction, proposeSigners);

// Extend it with a second transaction
const {extendTransaction, extendSigners} = await cryptid.extend(transactionAccount, tx2);
await cryptid.send(extendTransaction, extendSigners);

// Seal the transaction so that it cannot be modified
const {sealTransaction, sealSigners} = await cryptid.seal(transactionAccount);
await cryptid.send(sealTransaction, sealSigners);

// Execute the transaction
const {executeTransactions, executeSigners} = await cryptid.execute(transactionAccount);
executeTransactions.forEach(async (tx) => await cryptid.send(tx, executeSigners));
```

### CODE REVIEW

File: packages/client/core/src/lib/CryptidTransaction.ts [hunk 1]
```diff
@@ -119,43 +135,52 @@ export class CryptidTransaction {
...
       allAccounts,
       cryptidAccount
     );
-    const filteredAccountMetas = accountMetas.filter(
-      (a) =>
-        !namedAccounts
-          .map((account) => account.toBase58())
-          .includes(a.pubkey.toBase58())
-    );
+    const filteredAccountMetasFromInstructionsExcludingNamedAccounts =
```
Comment (rado0x54) [inline packages/client/core/src/lib/CryptidTransaction.ts:146]: I'm really trying to decide if that renaming makes it clearer or not :).

File: packages/client/core/src/service/cryptid.ts [hunk 2]
```diff
@@ -243,20 +261,25 @@ export class CryptidService {
...
+        account,
+        transactionAccountKeypair.publicKey,
+        "Propose"
+      );
+    }
 
     const unsignedProposeTransaction = await cryptidTransaction
-      .propose(this.program, transactionAccountKeypair.publicKey)
+      .propose(this.program, transactionAccountKeypair.publicKey, state)
       .signers(
```
Comment (rado0x54) [inline packages/client/core/src/service/cryptid.ts:276]: Good find. Strange that the double `.signers` call was not caught in a test. But I think we just don't have any middlewareResult with signers yet.

File: packages/tests/src/proposeExecute.ts [hunk 3]
```diff
@@ -49,10 +50,11 @@ didTestCases.forEach(({ didType, getDidAccount }) => {
       ) =>
         program.methods
           .proposeTransaction(
-            Buffer.from([]), // no controller chain,
+            [], // no controller chain,
```
Comment (rado0x54) [inline packages/tests/src/proposeExecute.ts:53]: I did not see any changes in the rust code around this and thought Vec<u8> needs to be a Buffer on the client side. Am I missing something?

File: programs/cryptid/src/instructions/propose_transaction.rs [hunk 4]
```diff
@@ -86,66 +85,52 @@ pub fn propose_transaction<'a, 'b, 'c, 'info>(
...
-        ctx.remaining_accounts.iter().map(|a| *a.key).collect();
+    ctx.accounts.transaction_account.accounts = all_accounts.iter().map(|a| *a.key).collect();
+
+    ctx.accounts.transaction_account.did = *ctx.accounts.did.key;
+    ctx.accounts.transaction_account.instructions = instructions;
+    ctx.accounts.transaction_account.cryptid_account = *ctx.accounts.cryptid_account.key;
+    ctx.accounts.transaction_account.approved_middleware = None;
+
+    // we cannot initiate a transaction in executed state.
+    require_neq!(
```
Comment (rado0x54) [inline programs/cryptid/src/instructions/propose_transaction.rs:128]: Ok I understand this is not part of this PR, but why do we need the state `ctx.accounts.transaction_account.state = TransactionState::Executed;` at all (and therefore check for it here)? The transaction_account get's closed right after successful execution, so i my mind the state (and the check is not needed?)

File: programs/cryptid/src/instructions/seal_transaction.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+/// Seals a transaction directly if all required keys sign
+pub fn seal_transaction<'a, 'b, 'c, 'info>(
+    ctx: Context<'a, 'b, 'c, 'info, SealTransaction<'info>>,
+    controller_authority_keys: Vec<Pubkey>,
+    cryptid_account_bump: u8,
+    cryptid_account_index: u32,
+    did_account_bump: u8,
+) -> Result<()> {
+    let controlling_did_accounts = ctx
```
Comment (rado0x54) [inline programs/cryptid/src/instructions/seal_transaction.rs]: We can use `get_cryptid_account_checked` here as well, right?

File: programs/cryptid/src/instructions/seal_transaction.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        mut,
+        has_one = cryptid_account @ CryptidError::WrongCryptidAccount,
+        // only transactions in "not ready" state can be sealed
+        constraint = transaction_account.state == TransactionState::NotReady @ CryptidError::InvalidTransactionState,
+    )]
+    pub transaction_account: Account<'info, TransactionAccount>,
+}
+
+/// Seals a transaction directly if all required keys sign
+pub fn seal_transaction<'a, 'b, 'c, 'info>(
```
Comment (rado0x54) [inline programs/cryptid/src/instructions/seal_transaction.rs]: Design question. Do we really need a dedicated instruction for sealing? Why not add another bool/u8 parameter to `extend_transaction` that indicates final/ready/sealed? You can extend with an empty `Vec<AbbreviatedInstructionData>` and just the bool. I think that would also simplify the code review.

**Review by rado0x54 — COMMENTED** [review]
Looks very good. Thanks Dan, I've left some comments that we can discuss in person as well.

File: packages/client/core/src/lib/CryptidTransaction.ts (hunk 1 above)
Comment (rado0x54) [reply packages/client/core/src/lib/CryptidTransaction.ts:146]: `ExludingNamedAccounts == RemainingAccounts`

Comment (dankelleher) [reply packages/client/core/src/lib/CryptidTransaction.ts:146]: renamed again

File: programs/cryptid/src/instructions/propose_transaction.rs (hunk 4 above)
Comment (dankelleher) [reply programs/cryptid/src/instructions/propose_transaction.rs:128]: as agreed, we keep it in for now.

File: programs/cryptid/src/instructions/seal_transaction.rs (hunk 5 above)
Comment (dankelleher) [reply programs/cryptid/src/instructions/seal_transaction.rs]: obsolete

File: programs/cryptid/src/instructions/seal_transaction.rs (hunk 6 above)
Comment (dankelleher) [reply programs/cryptid/src/instructions/seal_transaction.rs]: fixed

**Review by rado0x54 — APPROVED** [review]
